### ISSUE TITLE
Crash when disabling tcp server while connected

### ISSUE THREAD

**Issue by jeffrson** [dannagle/PacketSender#151]
#### What OS?

- [ X ] Windows
- [ ] Mac
- [ ] Linux (Which distro?)

#### Description of issue

I checked "Persistent TCP". When I disable TCP server by clicking the button (bottom TCP: port) while the seperate window with the persistent TCP connection is still open, PacketSender crashs.

**Comment by SurrealSky** [comment]
I also found this bug and tried to fix it. I hope the author can fix it as soon as possible.

**Comment by dannagle** [comment]
Yes, I was able to replicate this bug. It is being actively worked.

A fix didn't make it in to the release from a few days ago since what I need to do is pretty intense. For now, the workaround is to manually disconnect and close the pop-ups before disabling the server.

**Comment by bladerunner2020** [comment]
On Mac it also crashes.

**Comment by ProfessorTom** [comment]
@jeffrson Do you know QT or C++?

I tried to vibe-code a Connection Manager. While doing this work, I combined an `rm` command with the build command and I wiped out an hour's work. Consequently, I had to start from an in progress state via a Time Machine backup.

Here's [my WIP branch](https://github.com/ProfessorTom/PacketSender/tree/tomas/fix_crash_while_disabling_server)

I'm not saying it's right or it's production ready.

After I got to this point, I realized that the windows will still send packets even when the address scheme is wrong e.g. would send to IPV4 addresses when the main window had been toggled to IPV6. I tried fixing the last night and just went round and round in circles. Have a bunch of nasty, uncommitted, code that doesn't behave right on my machine.

Perhaps if I had some time to talk with @dannagle or got some guidance here, I could get a solution over the line.

Thoughts?

**Comment by bladerunner2020** [comment]
> [@bladerunner2020](https://github.com/bladerunner2020) [@jeffrson](https://github.com/jeffrson) [@SurrealSky](https://github.com/SurrealSky) Is this still an active bug?

PS v8.8.9, MacOS Apple Silicon 26.3 - the issue is still here.

**Comment by dannagle** [comment]
The problem is the hand-off of the socket to the new window. That is the wrong way to do it. You can blame me 10 years ago when I wrote it not quite understanding what I was doing.  I thought it would be faster, which it was not, and I made an architectural problem. 

The correct approach is to have the socket managed entirely by the main window and just create a data connection to/from the persistent window.  That way, when the TCP (or any server) is closed, that simple data connection is severed and not an orphaned thread.

**Comment by ProfessorTom** [comment]
@dannagle I don't want to blame anyone, I just want to fix this correctly.

I've got a PR up (#412 ) that adds unit tests to the project. Once you merge that in, my next task will be to unit test other low hanging fruit along the path to a fully tested connection manger.

There may be some side issues with creating more than one server and the servers not being terminated, but I'm not entirely sure.

C++ isn't my strong suit and prior to working on this project I had never worked with Qt before. I _think_ I'm starting to understand the current architecture, I just need more time to play with it.

Just like with the OS toggling between light and dark modes, I really want to solve this problem. Currently, I'm a bit out over my skies, but I feel this is within my ability given a bit more grind.

**Comment by ProfessorTom** [comment]
I'm going to try implementing this (or something very similar) with unit tests:

```MainWindow
   └── owns PacketNetwork (existing)
          └── owns ConnectionManager (new / renamed & upgraded)
                 └── owns std::vector<std::unique_ptr<Connection>>   // or QMap<quint64, std::unique_ptr<Connection>>
                         └── each Connection (RAII wrapper)
                                 ├── owns TCPThread* or Worker + QTcpSocket*
                                 ├── owns the socket
                                 └── emits dataReceived, stateChanged, error, etc.
```

**Comment by ProfessorTom** [comment]
WIP: https://github.com/ProfessorTom/PacketSender/tree/tomas/connection-raii-initial
